compiler/hie-core/src/Development/IDE/LSP/LanguageServer.hs

```diff
@@ -12,144 +11,102 @@ module Development.IDE.LSP.LanguageServer
-                    Just fp <- pure $ toNormalizedFilePath <$> uriToFilePath' uri
-                    handleDidCloseFile fp
-              | otherwise -> logWarning logger $ "Unknown scheme in URI: " <> textShow uri
-
-          _ -> logSeriousError logger
-                 $    "Invalid URI in DidCloseTextDocument: "
-                   <> textShow uri
-
-    DidSaveTextDocument _params ->
-      pure ()
-
-    UnknownNotification _method _params -> return ()
-  where
-    -- Note that the state changes here are not atomic.
-    -- When we have parallel compilation we could manage the state
-    -- changes in STM so that we can atomically change the state.
-    -- Internally it should be done via the IO oracle. See PROD-2808.
-    handleDidOpenFile (TextDocumentItem uri _ _ contents) = do
-        Just filePath <- pure $ toNormalizedFilePath <$> uriToFilePath' uri
-        onFileModified compilerH filePath (Just contents)
-        modifyFilesOfInterest compilerH (S.insert filePath)
-        logInfo logger $ "Opened text document: " <> textShow filePath
-
-    handleDidCloseFile filePath = do
-         logInfo logger $ "Closed text document: " <> textShow (fromNormalizedFilePath filePath)
-         onFileModified compilerH filePath Nothing
-         modifyFilesOfInterest compilerH (S.delete filePath)
-
--- | Manages the file store (caching compilation results and unsaved content).
-onFileModified
-    :: IdeState
-    -> NormalizedFilePath
-    -> Maybe T.Text
-    -> IO ()
-onFileModified service fp mbContents = do
-    logDebug (ideLogger service) $ "File modified " <> T.pack (show fp)
-    setBufferModified service fp mbContents
 
-------------------------------------------------------------------------
--- Server execution
-------------------------------------------------------------------------
 
 runLanguageServer
-    :: Logger
-    -> ((FromServerMessage -> IO ()) -> VFSHandle -> IO IdeState)
+    :: ((FromServerMessage -> IO ()) -> VFSHandle -> IO IdeState)
     -> IO ()
-runLanguageServer loggerH getIdeState = do
-    let getHandlers lspFuncs = do
-            compilerH <- getIdeState (sendFunc lspFuncs) (makeLSPVFSHandle lspFuncs)
-            pure $ Handlers (handleRequest loggerH compilerH) (handleNotification lspFuncs loggerH compilerH)
-    liftIO $ runServer loggerH getHandlers
+runLanguageServer getIdeState = do
+    -- Move stdout to another file descriptor and duplicate stderr
+    -- to stdout. This guards against stray prints from corrupting the JSON-RPC
+    -- message stream.
+    newStdout <- hDuplicate stdout
+    stderr `hDuplicateTo` stdout
+
+    -- Print out a single space to assert that the above redirection works.
+    -- This is interleaved with the logger, hence we just print a space here in
+    -- order not to mess up the output too much. Verified that this breaks
+    -- the language server tests without the redirection.
+    putStr " " >> hFlush stdout
+
+    clientMsgChan :: TChan AddItem <- newTChanIO
```

Comment:
Not really a new thing in your PR but it would be nice to have a comment on why we have this channel indirection. In the current state the answer is that we can only start handling messages once we have the `LspFuncs` but we need to specify the `Handlers` before that. It will probably also come in handy for implementing request cancellation.

Reply:
I'm not sure why we have this channel. I think it might be redundant. Either way, its just a copy of what we had before.